contributors/design-proposals/federated-deployment.md

```diff
@@ -0,0 +1,127 @@
+
+On the other hand FederatedReplicaSets should somehow reflect the status of the underlying cluster - like number of total replicas that are there so that kubectl rollout has proper data to work on. 
+
+<img src="federated-deployment-3.png" width="500"/>
+
+Federated nginx-701339712 gets all of the statistics as well as spec elements (like the total number of replicas) from ReplicaSets in the underlying clusters. Obviously none of the spec changes should be pushed back. With this change the whole thing would look like as a well-working “whole” while in fact Federated ReplicaSets would be “shadows” (or puppets) - their original functionality (creating rs in clusters and controlling the balancing) would be turned off. 
+
+## Flow
+
+As the above description may be a bit confusing let's describe what happens exactly when a user creates federated deployment:
+
+ 1. A user creates a Federated Deployment.
+ 2. Federated Deployment controller is notified about the new Federated Deployment.
+ 3. Federated Deployment controller creates Shadow Federated ReplicaSet.
+ 4. Federated Deployment controller creates Local Deployments in underlying clusters
+ 5. Local Deployments create Local ReplicaSets (in underlying clusters).
+ 6. Replicas are created by Local ReplicaSets.
+ 7. The status of Local ReplicaSets is updated.
+ 8. Federated ReplicaSet controller learn about changes in the underlying clusters.
+ 9. Shadow ReplicaSet status is updated based on Local ReplicaSet statuses.
+
+On update the situation looks similar:
+
+ 1. A user updates a Federated Deployment.
+ 2. Federated Deployment controller is notified about the update.
+ 3. Federated Deployment controller creates a new Shadow Federated ReplicaSet for the updated spec.
+ 4. Federated Deployment controller updates Local Deployments in underlying clusters
+ 5. Local Deployments create new Local ReplicaSets (in underlying clusters).
+ 6. Replicas are moved from the old Local ReplicaSet to the new LocalReplicaSet by Local Deployment controller. 
+ 7. The status of Local ReplicaSets is updated.
+ 8. Federated ReplicaSet controller learn about changes in the underlying clusters.
+ 9. Shadow ReplicaSet status is updated based on Local ReplicaSet statuses.
+
+
+## Implementation
+
+The implementation of Federated Deployment Controller is largely based on the existing Federated ReplicaSet Controller. The following changes however will be added (on top of obvious changes like “s/replicaset/deployment/”):
+
++ For each of the pod templates, Federated Deployment will make sure that shadow FederatedReplicaset with the name deployment_name + “-” + pod_template_hash is created, with the appropriate revision number.
```

Comment:
Hashing the podtemplate or any kind of api object is not reliable due to api changes. I will prepare a proposal during the weekend for switching to something stable.